src/compiler/declarationEmitter.ts

```diff
@@ -1237,14 +1237,23 @@ namespace ts {
             enclosingDeclaration = prevEnclosingDeclaration;
         }
 
-        function emitPropertyDeclaration(node: Declaration) {
-            if (hasDynamicName(node) && !resolver.isLateBound(node)) {
+        function hasNoncollidingLateBoundPropertyName(node: VariableDeclaration | PropertyDeclaration | PropertySignature | ParameterDeclaration) {
+            if (!hasModifier(node, ModifierFlags.Private)) {

```

Comment:
we want to keep the private in the type.. why remove late-bound names?

Reply:
If the name is impossible to posses outside of the local scope, then it is impossible to conflict with, ergo retaining the private holds no value.

Reply:
Plus, computed names like these were elided pre-dynamic-names anyway, so this is a conservative behavior.

Reply:
We do not know that.. you can have a function that returns you the same symbol.. the symbol can "escape" in multiple ways..

Reply:
In value-space, yes, the name can escape (all you'd need to do is alias the type or something); but it _can't_ as far as our type system in concerned - if the late bound name is exported then we'll write the private, since it is visible and we know the name can be referenced elsewhere, if it is not visible, then there's no public way to access the same name, ergo you can't conflict with it (or at least we have no way of checking).
For instance, if we allowed:
```ts
const _data = Symbol('data');

export function getDataSymbol() {
  return _data;
}

export class User {
    private [_data] : any;
};
```
`getDataSymbol` returns `symbol` (not a reference to `_data`'s `unique symbol`), even though it's easy to see that you could make a subclass that attempts to mangle the same symbol, we have no way of verifying that it is, in fact, the same symbol (since it didn't come from the same origin).

Reply:
And if you annotate `getDataSymbol` to explicitly return `typeof _data`, we'll happily tell you at that point in time that `_data` isn't exported and cannot be named.

Reply:
The problem is this case (simplified):
```ts
// a.ts
export const x = Symbol();
export function getX(): typeof x { return x; }

// b.ts
import { getX } from "./a";

const x = getX();

export class C {
  private [x]: number = 1;
}

// c.ts
import { getX } from "./a";
import { C } from "./b";

const x = getX();

export class D extends C {
  private [x]: string = "a"; // oops
}
```

Reply:
Hmm... I suppose the correct thing to check isn't visibility from the current declaration, but rather if the the name is reachable via the exports of its containing file at all. And then adding a visibility check to ensure an alias to it is visible in scope and complaining that it cannot be referenced if not. Meaning
```ts
import { getX } from "./a";
import { C } from "./b";

const x = getX();

export class D extends C {
  private [x]: string = "a"; // oops
}
```
is an error because `x` cannot be named in the declaration file (without modifying its shape, anyway), while
```ts
import { getX } from "./a";
import { C } from "./b";
import {x as xsym} from "./a";

const x = getX();

export class D extends C {
  private [x]: string = "a"; // oops
}
```
is fine since the name can be written by its alias from `a`.

Reply:
so why not just keep the private, and the const with the unique symbol, but just add a `export {};` to the file to make the scoping work correctly.. we talked about this before in the context of other declaration emit issues.